src/main/kotlin/com/terraformation/backend/tracking/db/ObservationResultsStore.kt

```diff
@@ -0,0 +1,346 @@
+          .convertFrom { results ->
+            results.map { record ->
+              val species = record[plantingZoneSpeciesMultiset]
+              val subzones = record[plantingSubzoneMultiset]
+
+              val isCompleted =
+                  subzones.all { subzone ->
+                    subzone.monitoringPlots.all { it.completedTime != null }
+                  }
+              val completedTime =
+                  if (isCompleted) {
+                    subzones.maxOf { subzone ->
+                      subzone.monitoringPlots.maxOf { it.completedTime!! }
+                    }
+                  } else {
+                    null
+                  }
+
+              val totalIdentifiedPlants =
+                  subzones.sumOf { subzone -> subzone.monitoringPlots.sumOf { it.totalPlants } }
+              val totalSpecies =
+                  subzones
+                      .flatMap { subzone ->
+                        subzone.monitoringPlots.flatMap { plot ->
+                          plot.species.map { it.speciesId }
+                        }
+                      }
+                      .distinct()
+                      .count()
+
+              val mortalityRate =
+                  if (totalIdentifiedPlants > 0) {
+                    species.sumOf { it.totalDead } * 100 / totalIdentifiedPlants
+                  } else {
+                    0
+                  }
+
+              // TODO: These need to count unidentified and existing plants too
+              val totalPlants = totalIdentifiedPlants
+              val plantingDensity =
+                  if (record[zoneFinishedPlantingField]) {
+                    (BigDecimal(totalPlants) / record[PLANTING_ZONES.AREA_HA.asNonNullable()])
+                        .toInt()
+                  } else {
+                    null
+                  }
+
+              ObservationPlantingZoneResultsPayload(
+                  completedTime = completedTime,
+                  mortalityRate = mortalityRate,
+                  plantingDensity = plantingDensity,
+                  plantingSubzones = subzones,
+                  plantingZoneId = record[PLANTING_ZONES.ID.asNonNullable()],
+                  species = species,
+                  totalPlants = totalPlants,
+                  totalSpecies = totalSpecies)
+            }
+          }
+
+  private fun fetchByCondition(condition: Condition): List<ObservationResultsPayload> {
+    return dslContext
+        .select(
+            OBSERVATIONS.ID,
+            OBSERVATIONS.PLANTING_SITE_ID,
+            OBSERVATIONS.START_DATE,
+            OBSERVATIONS.STATE_ID,
+            plantingZoneMultiset)
+        .from(OBSERVATIONS)
+        .where(condition)
+        .orderBy(OBSERVATIONS.ID)
+        .fetch { record ->
+          val zones = record[plantingZoneMultiset]
+
+          val isCompleted = zones.all { it.completedTime != null }
```

Comment:
For an upcoming observation, isCompleted is evaluating to true (zones is empty), do we need an isEmpty() check as well?  Ends up throwing an exception in the zones.maxOf (guarded by isCompleted).

Reply:
Good catch, thanks. This does indeed need to check for zones being non-empty.